gitriddance.1
.\" -*- nroff -*-
.TH gitriddance 1 "Commands"
.SH NAME
gitriddance \- closes all open issues and pull requests
.SH SYNOPSIS
.B gitriddance [comment]
.SH DESCRIPTION
.I gitriddance
closes all open issues and pull requests on GitHub. This is useful for
projects that have their own issue trackers, patch submission systems etc,
rather than relying on GitHub's, which many of us find to be clumsy,
slow, proprietary, and encouraging of drive-by pull requests of poor quality.
.PP
It should be run in a git repository that was cloned from GitHub. It
looks at the origin remote to find the repository on GitHub. All open
issues and pull requests will have a comment posted to them, and be closed.
.PP
The text of the comment is either passed as a command-line parameter,
or can be configured by setting core.gitriddance. For example:
.IP
git config core.gitriddance "Please submit patches to http://ikiwiki.info/todo/"
.PP
In order for gitriddance to log into GitHub, you need to set
GITHUB_OAUTH_TOKEN to an oath or personal access token.
.SH AUTHOR 
Joey Hess <id@joeyh.name>


CHANGELOG
github-backup (1.20200721) unstable; urgency=medium

  * Removed support for GITHUB_USER/GITHUB_PASSWORD since github will soon
    remove support for http basic auth. Use GITHUB_OAUTH_TOKEN instead.

 -- Joey Hess <id@joeyh.name>  Tue, 21 Jul 2020 23:24:03 -0400

github-backup (1.20191219) unstable; urgency=medium

  * Updated to use github-0.23 (0.24 needs larger changes)
  * Various updates to internal git and utility libraries shared
    with git-annex.
  * Remove support for building with old versions of network-uri and network
    before the split in 2.6.
  * stack.yaml: Updated to lts-13.29.

 -- Joey Hess <id@joeyh.name>  Thu, 19 Dec 2019 18:43:05 -0400

github-backup (1.20171126) unstable; urgency=medium

  * Updated to use github-0.18
  * Pull requests retreived with authentication again.
  * Build with -threaded, which fixes some issues with the non-threaded
    GHC RTS.
  * Removed dependency on MissingH, instead depending on split.
  * Various updates to internal git and utility libraries shared
    with git-annex.

 -- Joey Hess <id@joeyh.name>  Sun, 26 Nov 2017 15:33:35 -0400

github-backup (1.20170301) unstable; urgency=medium

  * No longer hosted on Github due to their horrible new TOS.
    New site: https://github-backup.branchable.com/

 -- Joey Hess <id@joeyh.name>  Wed, 01 Mar 2017 17:28:50 -0400

github-backup (1.20170221) unstable; urgency=medium

  * Updated to use github-0.15.
  * Due to a regression in github-0.15, the pull request list is retrieved
    without authentication.
  * Various updates to internal git and utility libraries shared
    with git-annex.

 -- Joey Hess <id@joeyh.name>  Tue, 21 Feb 2017 15:33:56 -0400

github-backup (1.20161118) unstable; urgency=medium

  * Fix build with recent versions of cabal.

 -- Joey Hess <id@joeyh.name>  Fri, 18 Nov 2016 12:58:19 -0400

github-backup (1.20161110) unstable; urgency=medium

  * Fix build with recent versions of the directory package.
    Thanks, James McCoy.
  * Various updates to internal git and utility libraries shared
    with git-annex.

 -- Joey Hess <id@joeyh.name>  Thu, 10 Nov 2016 11:55:44 -0400

github-backup (1.20160922) unstable; urgency=medium

  * Increase base bounds to 4.8 to get mapM_ etc that works on Vector.
  * Recommended build method is now to use stack, for more reliable builds.
  * Explicitly list modules and other files in github-backup.cabal.
  * The hackage .tar.gz now omits the debian directory and may omit other
    source files in the future; the purpose of that tarball is to make
    stack/cabal install work, not to be a complete source distribution.
    Use git repository to get complete source.
  * Various updates to internal git and utility libraries shared
    with git-annex.

 -- Joey Hess <id@joeyh.name>  Mon, 22 Aug 2016 23:57:16 -0400

github-backup (1.20160522) unstable; urgency=medium

  * Work around git weirdness in handling of relative path to GIT_INDEX_FILE
    when in a subdirectory of the repository.
  * Various updates to internal git and utility libraries shared
    with git-annex.
  * debian: Add lintian overrides for rpath.
  * debian: Bump standards-version.
  * Makefile: Pass LDFLAGS, CFLAGS, and CPPFLAGS through ghc and on to 
    ld, cc, and cpp. This lets the Debian package build with various
    hardening options. Although their benefit to a largely haskell program
    is unknown.

 -- Joey Hess <id@joeyh.name>  Sun, 22 May 2016 15:24:04 -0400

github-backup (1.20160511) unstable; urgency=medium

  * Fix build with directory-1.2.6.2.
  * github-backup.cabal: Add Setup-Depends.

 -- Joey Hess <id@joeyh.name>  Wed, 11 May 2016 13:12:12 -0400

github-backup (1.20160319) unstable; urgency=medium

  * Update to github-0.14.1.
    Building with old versions is not supported due to the massive API
    changes in this version of the github library.
  * Added support for oath tokens.

 -- Joey Hess <id@joeyh.name>  Sat, 19 Mar 2016 14:11:55 -0400

github-backup (1.20160207) unstable; urgency=medium

  * Switch to using https instead of git:// for a secure transport.
    Note that existing git:// remotes will continue to be used.
    Thanks, Robin Schneider.
  * Add upper bound on github dependency, as version 0.14 has large API
    changes and can't be used yet.
  * Various updates to internal git and utility libraries shared
    with git-annex.
  * Silence build warnings with ghc 7.10.

 -- Joey Hess <id@joeyh.name>  Sun, 07 Feb 2016 23:29:59 -0400

github-backup (1.20150807) unstable; urgency=medium

  * Added bash completion.
  * Add --no-forks flag that turns off backing up forks.
    Thanks, Phil Ruffwind.
  * Avoid nonzero exit due to temporary failures that can be retried next
    time github-backup runs, so that it can be used in a cron job with eg,
    chronic.

 -- Joey Hess <id@joeyh.name>  Fri, 07 Aug 2015 12:50:02 -0400

github-backup (1.20150618) unstable; urgency=medium

  * Fix broken configure script.

 -- Joey Hess <id@joeyh.name>  Thu, 18 Jun 2015 16:56:41 -0400

github-backup (1.20150617) unstable; urgency=medium

  * Add missing build deps for windows to cabal file.
    Thanks, Jeff Segal.
  * Various updates to internal git and utility libraries shared
    with git-annex.

 -- Joey Hess <id@joeyh.name>  Wed, 17 Jun 2015 14:44:14 -0400

github-backup (1.20150106) unstable; urgency=medium

  * Fix build with process 1.2.1.0.
  * Various updates to internal git and utility libraries shared
    with git-annex.

 -- Joey Hess <id@joeyh.name>  Tue, 06 Jan 2015 19:08:08 -0400

github-backup (1.20141222) unstable; urgency=medium

  * Added gitriddance(1), a utility to close all issues and pull requests,
    for repos that don't want to be bothered with GitHub's proprietary
    issue tracker.
  * gitriddance depends on github 0.13.1, which has bug fixes
    for posting comments.
  * Various updates to internal git and utility libraries shared
    with git-annex.

 -- Joey Hess <id@joeyh.name>  Mon, 22 Dec 2014 15:30:08 -0400


gitriddance.hs
{- gitriddance - close all open issues and pull requests
 -
 - Copyright 2014 Joey Hess <joey@kitenet.net>
 -
 - Licensed under the GNU GPL version 3 or higher.
 -}

module Main where

import qualified GitHub.Endpoints.Repos as Github
import qualified GitHub.Endpoints.Issues as Github
import qualified GitHub.Endpoints.Issues.Comments as Github
import qualified GitHub.Data.Id as Github
import System.Environment
import Data.String
import qualified Data.Text as T

import Common
import qualified Git
import qualified Git.Construct
import qualified Git.Config
import Github.GetAuth
import Github.EnumRepos

main :: IO ()
main = do
	auth <- fromMaybe (error "Must set GITHUB_USER and GITHUB_PASSWORD")
		<$> getAuth
	r <- maybe (error "not in a git repository") Git.Config.read
		=<< Git.Construct.fromCwd
	rs <- Git.Construct.fromRemotes r
	msg <- maybe (getMsg r) id <$> (headMaybe <$> getArgs)
	case map snd $ filter (isOriginRemote . fst) $ gitHubPairs rs of
		[] -> error "origin does not seem to be a github repository"
		[origin] -> closeall auth origin msg
		_ -> error "somehow found multiple origin repos; this should be impossible!"

getMsg :: Git.Repo -> String
getMsg r = fromMaybe (error "core.gitriddance needs to be set to a message to use when closing issues/pull requests (or pass the message on the command line)")
	(Git.Config.getMaybe "core.gitriddance" r)

{- Identify the origin remote; we don't want to affect any
 - other remotes that might be on github. -}
isOriginRemote :: Git.Repo -> Bool
isOriginRemote rmt = Git.remoteName rmt == Just "origin"

closeall :: Github.Auth -> GithubUserRepo -> String -> IO ()
closeall auth (GithubUserRepo user repo) msg =
	either (oops "getting issue list") (mapM_ close)
		=<< Github.issuesForRepo' (Just auth) (fromString user) (fromString repo) Github.stateOpen
  where
	oops action err = error $ "failed " ++ action ++ ": " ++ show err
	close issue = do
		let i = Github.issueNumber issue
		putStrLn $ "closing issue: " ++ T.unpack (Github.issueTitle issue)
		either (oops "posting comment") (const $ return ())
			=<< Github.createComment auth (fromString user) (fromString repo) i (T.pack msg)
		either (oops "closing issue/pull") (const $ return ())
			=<< Github.editIssue auth (fromString user) (fromString repo) (Github.mkId (Github.Id 0) (Github.unIssueNumber i))
				(Github.editOfIssue { Github.editIssueState = Just Github.StateClosed } )
